lib/modules/platform/gitlab/pr-cache.ts

```diff
@@ -0,0 +1,174 @@
+import type { GitLabMergeRequest, GitlabPr, GitlabPrCacheData } from './types';
+import { prInfo } from './utils';
+
+export class GitlabPrCache {
+  private items: GitlabPr[] = [];
+  private cache: GitlabPrCacheData;
+
+  private constructor(
+    private repo: string,
+    author: string | null,
+    private ignorePrAuthor: boolean,
+  ) {
+    const repoCache = getCache();
+    repoCache.platform ??= {};
+    repoCache.platform.gitlab ??= {};
+
+    let pullRequestCache = repoCache.platform.gitlab
+      .pullRequestsCache as GitlabPrCacheData;
+    if (!pullRequestCache) {
+      logger.debug('Initializing new PR cache at repository cache');
+      pullRequestCache = {
+        items: {},
+        updated_at: null,
+        author,
+      };
+    } else if (pullRequestCache.author !== author) {
+      logger.debug('Resetting PR cache because authors do not match');
+      pullRequestCache = {
+        items: {},
+        updated_at: null,
+        author,
+      };
+    }
+    repoCache.platform.gitlab.pullRequestsCache = pullRequestCache;
+    this.cache = pullRequestCache;
+    this.updateItems();
+  }
+
+  private static async init(
+    http: GitlabHttp,
+    repo: string,
+    author: string | null,
+    ignorePrAuthor: boolean,
+  ): Promise<GitlabPrCache> {
+    const res = new GitlabPrCache(repo, author, ignorePrAuthor);
+    const isSynced = memCache.get<true | undefined>('gitlab-pr-cache-synced');
+
+    if (!isSynced) {
+      await res.sync(http);
+      memCache.set('gitlab-pr-cache-synced', true);
+    }
+
+    return res;
+  }
+
+  private getPrs(): GitlabPr[] {
+    return this.items;
+  }
+
+  static async getPrs(
+    http: GitlabHttp,
+    repo: string,
+    author: string | null,
+    ignorePrAuthor: boolean,
+  ): Promise<GitlabPr[]> {
+    const prCache = await GitlabPrCache.init(
+      http,
+      repo,
+      author,
+      ignorePrAuthor,
+    );
+    return prCache.getPrs();
+  }
+
+  private setPr(pr: GitlabPr): void {
+    logger.debug(`Adding PR #${pr.number} to the PR cache`);
+    this.cache.items[pr.number] = pr;
+    this.updateItems();
+  }
+
+  static async setPr(
+    http: GitlabHttp,
+    repo: string,
+    author: string | null,
+    item: GitlabPr,
+    ignorePrAuthor: boolean,
+  ): Promise<void> {
+    const prCache = await GitlabPrCache.init(
+      http,
+      repo,
+      author,
+      ignorePrAuthor,
+    );
+    prCache.setPr(item);
+  }
+
+  private reconcile(rawItems: GitLabMergeRequest[]): boolean {
+    const { items: oldItems } = this.cache;
+    let { updated_at } = this.cache;
+
+    let needNextPage = true;
+
+    for (const rawItem of rawItems) {
+      const id = rawItem.iid;
+
+      const oldItem = oldItems[id];
+      const newItem = prInfo(rawItem);
+
+      const itemNewTime = DateTime.fromISO(rawItem.updated_at);
+
+      if (dequal(oldItem, newItem)) {
+        needNextPage = false;
+        continue;
+      }
```

Comment:
do we really need to do this? is this for the case we need new pr properties?

Reply:
It is the same in other implementations, maybe we can refactor this together in followup

Reply:
ok, it would be a huge performance improvement because comaring a lot of those big objects is more expensive than comparing a date 🙃